File: client/src/types/TransactionRow.ts
import type { RowType } from "./RowType";
import type { RowCompute } from "./CurComputeTypes";

export interface TransactionRow {
  // state management
  id: string;
  parentId: string | null;
  isSubRow: boolean;
  finalized: boolean;
  // columns
  date: string;
  rowType: RowType;
  inAmount: number | null;
  inCurrency: string | null;
  outAmount: number | null;
  outCurrency: string | null;
  feeAmount: number | "auto" | null;
  feeCurrency: string | null;
  usdValue: number | null;
  network: string;
  tags: string[];
  note: string;
  // css
  rowClass: string;
  // computed
  sortGroup: number;
  compute?: RowCompute;
}

export enum RowClassEnum {
  NOT_FINALIZED = "row-not-finalized",
  SUB_ROW = "row-sub",
  REGULAR = "row-regular",
}

File: server/serverType/TransactionRow.py
from dataclasses import dataclass
from pydantic import BaseModel, Field, ConfigDict
from typing import List, Optional, Union, Literal
from serverType.RowType import RowType


class TransactionRow(BaseModel):
    # state management
    id: str
    parentId: Optional[str] = None
    isSubRow: bool
    # columns
    date: str
    rowType: RowType
    inAmount: Optional[Union[float, None]]
    inCurrency: Optional[str]
    outAmount: Optional[Union[float, None]]
    outCurrency: Optional[str]
    feeAmount: Union[float, str, None]  # Can be "auto" or a number
    feeCurrency: Optional[str]
    usdValue: Optional[Union[float, None]]
    network: str
    tags: List[str]
    note: Optional[str]

    # config to ignore extra fields
    model_config = ConfigDict(extra='ignore')
